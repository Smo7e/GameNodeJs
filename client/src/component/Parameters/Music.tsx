import React, { memo, useContext, useEffect, useRef, useState } from "react"; 
import { MediatorContext } from "../../App"; 
import mp3 from './mp3.mp3'; 
 
const Music: React.FC = memo(() => { 
    const mediator = useContext(MediatorContext); 
    const audioRef = useRef<HTMLAudioElement | null>(null); 
    const [audio, setAudio] = useState(new Audio(mp3)); 
    audio.loop = true; 
 
    const handlePlay = () => { 
        setAudio(new Audio(mp3)); 
    }; 
    const handleStop = () => { 
        audio.pause(); 
    }; 
 
    const handleVolumeChange = (data: { volume: number }) => { 
        audio.volume = data.volume; 
    }; 
    useEffect(() => { 
        const { PLAY_MUSIC, STOP_MUSIC, SET_MUSIC_VOLUME } = mediator.getEventTypes(); 
 
        audioRef.current = audio; 

        audio.play(); 
 
        mediator.subscribe(PLAY_MUSIC, handlePlay); 
        mediator.subscribe(STOP_MUSIC, handleStop); 
        mediator.subscribe(SET_MUSIC_VOLUME, handleVolumeChange); 
 
        return () => { 
            mediator.unsubscribe(PLAY_MUSIC, handlePlay); 
            mediator.unsubscribe(STOP_MUSIC, handleStop); 
            audio.pause(); 
        }; 
    }); 
 
    return null; 
}); 
export default Music;
