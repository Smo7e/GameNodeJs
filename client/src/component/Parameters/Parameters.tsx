import React, { useState, useEffect, useContext } from "react";
import { EPAGES, MediatorContext } from "../../App";
import Music from "./Music";
import "./Parameters.css";

interface IParameters {
    epages: Function;
}

const Parameters: React.FC<IParameters> = ({ epages }) => {
    const mediator = useContext(MediatorContext);
  const [volume, setVolume] = useState(() => {
    const savedVolume = localStorage.getItem("volume");
    return savedVolume ? Number(savedVolume) : 60;
  });

  const [musicVolume, setMusicVolume] = useState(() => {
    const savedMusicVolume = localStorage.getItem("musicVolume");
    return savedMusicVolume ? Number(savedMusicVolume) : 77;
  });

  const [fullscreen, setFullscreen] = useState(false);

  const handleVolumeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newVolume = Number(event.target.value);
    setVolume(newVolume);
    localStorage.setItem("volume", newVolume.toString());
  };

  const handleMusicVolumeChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const newMusicVolume = Number(event.target.value);
    setMusicVolume(newMusicVolume);
    localStorage.setItem("musicVolume", newMusicVolume.toString());
    mediator.call("SET_MUSIC_VOLUME", { volume: newMusicVolume / 100 });
  };

  const handleFullscreenChange = () => {
    if (document.documentElement.requestFullscreen) {
      if (!document.fullscreenElement) {
        document.documentElement.requestFullscreen();
      } else {
        if (document.exitFullscreen) {
          document.exitFullscreen();
        }
      }
      setFullscreen(!fullscreen);
    }
  };

  useEffect(() => {
    const handleFullscreenChange = () => {
      setFullscreen(!!document.fullscreenElement);
    };
    document.addEventListener("fullscreenchange", handleFullscreenChange);

    return () => {
      document.removeEventListener("fullscreenchange", handleFullscreenChange);
    };
  }, []);

  return (
    <div className="container-parameters">
      <button onClick={() => epages(EPAGES.MENU)} className="arrow-parameters"></button>
      <div className="rectangle">
        <div className="inner-rectangle">
          <div className="line"></div>
          <div className="text-container">
            <div className="text">Громкость звуков</div>
            <div className="volume-line">
              <input
                type="range"
                min={0}
                max={100}
                value={volume}
                onChange={handleVolumeChange}
                className="volume-slider"
                style={{
                  background: `linear-gradient(to right, black 0%, black ${volume}%, white ${volume}%, white 100%)`
                }}
              />
            </div>
            <div className="numbersZ">{volume}</div>
            <div className="text">Громкость музыки</div>
            <div>
              <input
                type="range"
                min={0}
                max={100}
                value={musicVolume}
                onChange={handleMusicVolumeChange}
                className="lineM"
                style={{
                  background: `linear-gradient(to right, black 0%, black ${musicVolume}%, white ${musicVolume}%, white 100%)`
                }}
              />
            </div>
            <div className="numbersM">{musicVolume}</div>
            <div className="text">Полноэкранный режим</div>
            <div className="second-rectangle">
              <div
                className={`left-side ${!fullscreen ? "active" : ""}`}
                onClick={handleFullscreenChange}
              >
                Выкл
              </div>
              <div
                className={`right-side ${fullscreen ? "active" : ""}`}
                onClick={handleFullscreenChange}
              >
                Вкл
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
export default Parameters;