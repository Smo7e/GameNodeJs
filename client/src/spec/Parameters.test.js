import React from "react";
import { render, screen } from "@testing-library/react";
import Parameters from "../component/Parameters/Parameters";


test("Parameters snapshot", () => {
    const { asFragment } = render(
                <Parameters epages={() => {}} />
    );
    expect(asFragment()).toMatchSnapshot();
});
