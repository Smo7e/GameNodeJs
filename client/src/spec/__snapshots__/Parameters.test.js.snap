// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parameters snapshot 1`] = `
<DocumentFragment>
  <div
    class="container-parameters"
  >
    <button
      class="arrow-parameters"
    />
    <div
      class="rectangle"
    >
      <div
        class="inner-rectangle"
      >
        <div
          class="line"
        />
        <div
          class="text-container"
        >
          <div
            class="text"
          >
            Громкость звуков
          </div>
          <div
            class="volume-line"
          >
            <input
              class="volume-slider"
              max="100"
              min="0"
              type="range"
              value="60"
            />
          </div>
          <div
            class="numbersZ"
          >
            60
          </div>
          <div
            class="text"
          >
            Громкость музыки
          </div>
          <div>
            <input
              class="lineM"
              max="100"
              min="0"
              type="range"
              value="77"
            />
          </div>
          <div
            class="numbersM"
          >
            77
          </div>
          <div
            class="text"
          >
            Полноэкранный режим
          </div>
          <div
            class="second-rectangle"
          >
            <div
              class="left-side active"
            >
              Выкл
            </div>
            <div
              class="right-side "
            >
              Вкл
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
